BlueprintUI/Sources/Layout/LayoutWriter.swift

```diff
@@ -0,0 +1,227 @@
+            }
+        }
+    }
+    
+    /// Provides the relevant information about the context in which the layout is occurring.
+    public struct Context {
+        
+        /// The size constraint the layout is occurring in.
+        public var size : SizeConstraint
+    }
+    
+    /// Controls the sizing calculation of the custom layout.
+    public enum Sizing : Equatable {
+        
+        /// Ensures that the final size of element is large enough to fit all children, starting from (0,0).
+        ///
+        /// Negative origins of rects are not considered in this calculation. If you have the following layout:
+        /// ```
+        ///  ┌──────┐
+        ///  │      ├─────────┐
+        ///  │      │*********│
+        ///  └─┬────┘**┌──────┤
+        ///    │*******│      │
+        ///    │*******│      │
+        /// ┌──┴───┐***│      │
+        /// │      │***│      │
+        /// │      │***└──────┤
+        /// └──────┴──────────┘
+        /// ```
+        /// The large rect will be the calculated size / bounds of the layout, starting at (0,0). Any rects with
+        /// negative origins will overhang the layout to the top or left, respectively.
+        case unionOfChildren
+        
+        /// Fixes the layout size to the provided size. Children are positioned within this size, starting at (0,0)
+        /// Any rects with negative origins will overhang the layout to the top or left, respectively.
+        case fixed(CGSize)
+        
+        /// Measures the size of the content within the builder.
+        func measure(with builder : Builder) -> CGSize {
+            switch self {
+            case .unionOfChildren:
+                return CGSize(
+                    width: builder.children.reduce(0.0) { width, child in
+                        max(width, child.frame.maxX)
+                    },
+                    height: builder.children.reduce(0.0) { height, child in
+                        max(height, child.frame.maxY)
+                    }
+                )
+            case .fixed(let size):
+                return size
+            }
+        }
+    }
+    
+    /// A child of the custom layout, providing its frame and element.
+    public struct Child {
+        
+        /// The frame of the element in the coordinate space of the custom layout.
+        public var frame : CGRect
+        
+        /// The element to be displayed.
+        public var element : Element
+        
+        /// Creates a new child element.
+        public init(
+            frame : CGRect,
+            element : Element
+        ) {
+            self.frame = frame
+            self.element = element
+        }
+    }
+}
+
+
+extension LayoutWriter {
+    
+    private struct InnerElement : Element {
+        var builder : Builder
+        
+        // MARK: Element
+        
+        var content: ElementContent {
+            ElementContent(layout: Layout(builder: self.builder)) { builder in
```

Comment:
Ohhhhh right, good call

Reply:
Actually, hmm, I might still need it; lemme check if the various ElementContent inits do what I need (need the size).

Reply:
Though I guess if I re-build during measure and layout, that won't matter, which I think happens anyway

Reply:
oh duh, the one with measureFunction as a param...

Reply:
Oh wait that only returns a CGSize (also duh...), so I think I do need the inner element – since I basically need to lazily generate the content, depending on the size passed by the parent.

Reply:
Ok this works now, thanks for coming to my stream of consciousness.

Reply:
Ok actuallllly wait this won't work; since the children of the layout writer might change based on the provided layout size, so therefore we can't build the layout and provide the children with 

```
        ElementContent(layout: Layout(build: self.build)) { builder in
            builder.add(element: ...)
        }
```

Since the children aren't actually know at that point.